man/sdhash.1
.\" Automatically generated by Pod::Man 2.25 (Pod::Simple 3.16)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "SDHASH 1"
.TH SDHASH 1 "2012-12-07" "" ""
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
sdhash \- tool for generating and comparing similarity digests
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBsdhash\fR [options] [mode] <files>
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBsdhash\fR is a tool for generating, comparing, and searching similarity digest 
hashes.  When comparing these hashes, it gives a confidence level as to how similar
two objects are to each other.
.PP
\&\fBsdhash\fR uses parallel processing to efficiently process large quantities of data
and provides options for sampling to speed up queries.
.PP
\&\fBsdhash\fR has three modes: generation, comparison, and searching.
.SH "EXAMPLES"
.IX Header "EXAMPLES"
\&\fBsdhash\fR file1 file2 file3...
.PP
Generation mode produces digests for the files given on the command line.
.PP
\&\fBsdhash\fR \-\-compare set1.sdbf set2.sdbf
.PP
In comparison mode the contents of two sets of sdbf digests are compared
to produce a confidence score between \-1 and 100.
.PP
\&\fBsdhash\fR \-\-search\-deep \-\-index\-dir=references/ file1 file2 file3
.PP
Searching a reference set gives how many features in the hashed files
have matches in the indexed digests.  It is often faster than full comparisons.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-\-verbose\fR" 4
.IX Item "--verbose"
Displays (lots of) helpful output on every stage of the hashing process.
.IP "\fB\-f\fR, \fB\-\-hash\-list\fR <filelist.txt>" 4
.IX Item "-f, --hash-list <filelist.txt>"
Reads list of files to be hashed from given file.  File should be unix formatted 
with one filename per line.  Applies to generate mode and generate and compare mode.
.IP "\fB\-\fR" 4
.IX Item "-"
Reads input to be hashed from stdin.  By default uses 16kb blocks and supports
threading.
.IP "\fB\-r\fR, \fB\-\-deep\fR" 4
.IX Item "-r, --deep"
Searches any directories given at the command line for files to hash, recursively.
.IP "\fB\-o\fR, \fB\-\-output\fR <filename>" 4
.IX Item "-o, --output <filename>"
When creating hashes, sends output to named sdbf files and indexes.
.IP "\fB\-b\fR, \fB\-\-block\-size\fR <N>" 4
.IX Item "-b, --block-size <N>"
Configures block mode during hash generation.  To disable block mode, use \-b 0. 
Block mode is turned on for any files larger than 16MB, using 16KB blocks.
We suggest using \-b 4 for memory dumps in 4KB blocks.
.IP "\fB\-p\fR, \fB\-\-thread\-count\fR <N>" 4
.IX Item "-p, --thread-count <N>"
Use N threads.  Set this to the number of hardware threads available
on the local machine.  Available for all modes.
.IP "\fB\-t\fR, \fB\-\-threshold\fR <\-1\-100>" 4
.IX Item "-t, --threshold <-1-100>"
Sets a confidence threshold for comparisions.  Use this to filter results
to reduce false positives.  Used with comparison and query modes. 
Use '\-1' to display all comparisons.
.RS 4
.Sp
.RS 4
\&\fBsdhash\fR \fB\-\-threshold\fR 10 \fB\-\-compare\fR output.sdbf
.Sp
\&\fBsdhash\fR \fB\-\-threshold\fR '\-1' \fB\-\-compare\fR output.sdbf
.RE
.RE
.RS 4
.RE
.IP "\fB\-s\fR, \fB\-\-sample\-size\fR <1\-16>" 4
.IX Item "-s, --sample-size <1-16>"
Uses a sampling of filters to perform queries, which speeds up
query results significantly.  Used for query-mode only.
.IP "\fB\-n\fR, \fB\-\-name\fR <some name>" 4
.IX Item "-n, --name <some name>"
Sets the hash name when hashing from stdin.
.IP "\fB\-z\fR, \fB\-\-segment\-size\fR <\s-1NNN\s0>" 4
.IX Item "-z, --segment-size <NNN>"
Sets the default segment-size in \s-1MB\s0 to split files/streams into prior to hashing.  
Default is 128MB chunks.  Disable segmentation by passing 0 as the argument.
.IP "\fB\-m\fR, \fB\-\-heat\-map\fR" 4
.IX Item "-m, --heat-map"
Displays a heat-map of matches while in comparision or query mode. Heat map is a debugging 
or sanity check feature which works for single-threaded comparisons only.
.IP "\fB\-w\fR, \fB\-\-warnings\fR" 4
.IX Item "-w, --warnings"
Displays warnings.
.IP "\fB\-\-validate\fR <files>" 4
.IX Item "--validate <files>"
Checks if the given sdbf files are valid.
.IP "\fB\-C\fR, \fB\-\-config\fR <sdhash.cfg>" 4
.IX Item "-C, --config <sdhash.cfg>"
Reads options from configuration file.  Any option from the usage statement
can be set in the configuration file.
.IP "\fBsdhash.cfg\fR" 4
.IX Item "sdhash.cfg"
Options to sdhash may be saved in sdhash.cfg in the format of long\-option=value.
.Sp
For example:
.RS 4
.Sp

ChangeLog
New in version 3.3

Major bugfix involving self-to-self block-based comparison scoring,
we are now evenly discounting unfilled filters.

Added --separator change option.

Clarified indexing options/formats.

Enabled support for multiple lists-of-files to hash.

Now passing -Wall on gcc on Linux in main program.

--------
New in version 3.2.5

Significant speedup of CPU comparisons via openMP threads.  Now 
requiring openMP. 

Program will automatically use the maximum number of cores/threads
as given by openMP, on generation and comparison.
--------
New in version 3.2

CUDA-accelerated GPU comparison program available - sdhash-gpu. 

Changes to indexing to auto-generate indexes of larger, more 
usable size (640MB) to search much more information.

Small change to bloom filter fullness threshold for comparison,
we now require 16 elements instead of 6.
--------
New in version 3.1

New comparison code using the POPCNT instruction.  Marked speed improvements 
on intel, cache performance improvements on AMD.

Refined the searching/index components, including auto-generation of reference
set from large quantity of small files, search shortcuts, etc.  Indexes are
now stored compressed on disk using lz4.

More additions to --verbose as progress indicators.

Overhaul of sdhash-cli options, matching sdhash as closely as possible and much 
more flexible.  Also removing last traces of posix from sdhash-cli for inclusion 
into windows port.
--------
New in version 3.0

sdbf_set indexing method, to be generated with --index while hashing
and searched with --search-index while hashing.

temp-bugfix: some speedup for very small reference data
bugfix: segmenting was seeing too-large final segment on windows
--------
New in version 2.3

Web interface using python/jquery -- talks to thrift
--validate option
tons of --verbose debugging output
New thrift-server sources directory
bugfix: for invalid sdbf file reading no longer crashes
bugfix: windows recursive search filesystem permissions crash fixed

--------
New in version 2.2

Very beta native win32/win64 port
Threads updated to use boost::thread instead of pthreads directly
Program options updated to support all command line options in sdhash.cfg
Filesystem access now primarily c++/boost not posix
Thrift client/server now supports asynchronous mode comparisions
output file option -o --output 

--------

New in version 2.1

recursive directory hashing with --deep -r option
sdhash supports input from stdin
argument --segment-size -z to customize read-in segment size
argument --name -n to rename stdin hashes
new client/server programs using Thrift
api updates to use const and clearer naming conventions
api updated with formal 'set' class
sdhash auto-switching to block mode when files are >16MB
new base64 encoding library

--------


New in version 2.0

gnu-style long-options for standalone and client programs
configuration file capability for standalone sdhash sdhash.cfg
fixed: sdhash buffers large files instead of memory mapping the entire thing
sdhash server program sdhashd, and sdhash-cli client program.
Client and server should be considered beta software.
sdhash-dd program removed and changed to -b option for block size
sdbf version upgraded to add 'original size' into the header information
"view" option to create subsets of sets matching an expression
added: Boost libraries for regular expressions
fixed: import as file list now can use DD mode
fixed: sampling now non-destructive
swig-python bindings now exist with a sample program
Now large block hashes are "streamed" in chunks instead of as one giant hash.

---------

New in version 1.8

-i file.txt option to generate hashes from a file list. 
conversion to C++ 
beta libsdbf API 
fixed: off-by-one error on last line of query file

---------

New in version 1.7

-c query.sdbf target.sdbf option
-s sample option
Support for cygwin
manpage for sdhash
make install supported
fixed: spaces in filenames now readable on input sdbf

---------


sdhash-src/sdhash.cc
/**
 * sdhash: Command-line interface for file hashing
 * authors: Vassil Roussev, Candice Quates
 */

#include "../sdbf/sdbf_class.h"
#include "../sdbf/sdbf_defines.h"
#include "../sdbf/sdbf_set.h"
#include "sdhash_threads.h"
#include "sdhash.h"
#include "version.h"

#include "boost/filesystem.hpp"
#include "boost/program_options.hpp"
#include "boost/lexical_cast.hpp"

#include <fstream>

namespace fs = boost::filesystem;
namespace po = boost::program_options;

std::string read_file(const char* fname) {
    int length;
    char * buffer;

    ifstream is;
    is.open (fname, ios::binary );

    // get length of file:
    is.seekg (0, ios::end);
    length = is.tellg();
    is.seekg (0, ios::beg);

    // allocate memory:
    buffer = new char [length];

    // read data as a block:
    is.read (buffer,length);

    is.close();

    std::string r(buffer,length);
    delete [] buffer;
    return r;
}

double utcsecond()
{
    struct timeval tv;
    gettimeofday( &tv, NULL );
    return (double)(tv.tv_sec) + ((double)(tv.tv_usec))/1000000.0f;
}



// Global parameter configuration defaults
sdbf_parameters_t sdbf_sys = {
    1,               // threads
    64,              // entr_win_size
    256,             // BF size
    4*KB,            // block_size
    64,              // pop_win_size
    16,              // threshold
    _MAX_ELEM_COUNT, // max_elem
    1,               // output_threshold
    FLAG_OFF,        // warnings
    -1,              // dd block size
    0,              // sample size off
    0,              // verbose mode off
    128*MB,         // segment size
    NULL             // optional filename
};


/** sdhash program main
*/
int main( int argc, char **argv) {
    uint32_t  i, j, k, file_cnt;
    int rcf;
    time_t hash_start;
    time_t hash_end; 
    string config_file;
    string listingfile;
    string input_name;
    string output_name;
    string segment_size;
    string idx_size;
    string idx_dir; // where to find indexes
    uint32_t index_size = 16*MB; // default?
    vector<string> inputlist;
    po::variables_map vm;
    po::options_description config("Configuration");
    try {
        // Declare a group of options that will be 
        // allowed both on command line and in
        // config file
        config.add_options()
                ("config-file,C", po::value<string>(&config_file)->default_value("sdhash.cfg"), "name of config file")
                ("hash-list,f",po::value<std::string>(&listingfile),"generate SDBFs from list of filenames")
                ("deep,r", "generate SDBFs from directories and files")
                ("gen-compare,g", "generate SDBFs and compare all pairs")
                ("compare,c","compare all pairs in SDBF file, or compare two SDBF files to each other")
                ("benchmark,B","compare two SDBF files to each other, and do a benchmark")
                ("threshold,t",po::value<int32_t>(&sdbf_sys.output_threshold)->default_value(1),"only show results >=threshold")
                ("block-size,b",po::value<int32_t>(&sdbf_sys.dd_block_size),"hashes input files in nKB blocks")
                ("threads,p",po::value<uint32_t>(&sdbf_sys.thread_cnt)->default_value(1),"compute threads to use")
                ("sample-size,s",po::value<uint32_t>(&sdbf_sys.sample_size)->default_value(0),"sample N filters for comparisons")
                ("segment-size,z",po::value<std::string>(&segment_size),"break files into segments before hashing")
                ("name,n",po::value<std::string>(&input_name),"set SDBF name for stdin mode")
                ("output,o",po::value<std::string>(&output_name),"set output filename")
                ("heat-map,m", "show a heat map of BF matches")
                ("validate","parse SDBF file to check if it is valid")
                ("index","generate indexes while hashing")
                ("index-dir",po::value<std::string>(&idx_dir),"compare against reference indexes")
                ("search-all","match at file level, all matching sets")
                ("search-first","match at file level, first match")
                ("basename","print set matches with only base filenames")
                ("warnings,w","turn on warnings")
                ("verbose","debugging and progress output")
                ("version","show version info")
                ("help,h","produce help message")
            ;

        // options that do not need to be in help message
        po::options_description hidden("Hidden");
        hidden.add_options()
            ("input-files", po::value<vector<std::string> >(&inputlist), "input file") 
            ;

        po::options_description cmdline_options;
        cmdline_options.add(config).add(hidden);

        po::options_description config_file_options;
        config_file_options.add(config);

        // setup for list of files on command line
        po::positional_options_description p;
        p.add("input-files", -1);
        
        store(po::command_line_parser(argc, argv).
              options(cmdline_options).positional(p).run(), vm);
        notify(vm);
        
        ifstream ifs(config_file.c_str());
        if (ifs)
        {
            store(parse_config_file(ifs, config_file_options), vm);
            notify(vm);
        }
    
        if (vm.count("help")) {
            cout << VERSION_INFO << ", rev " << REVISION << endl;
            cout << "Usage: sdhash <options> <source files>|<hash files>"<< endl;
            cout << config << endl;
            return 0;
        }

        if (vm.count("version")) {
            cout << VERSION_INFO << ", rev " << REVISION << endl;
            cout << "       http://sdhash.org, license Apache v2.0" << endl;
            return 0;
        }
        if (vm.count("warnings")) {
            sdbf_sys.warnings = 1;
        }
        if (vm.count("verbose")) {
            sdbf_sys.warnings = 1;
            sdbf_sys.verbose = 1;
        }
        if (vm.count("segment-size")) {
            sdbf_sys.segment_size = (boost::lexical_cast<uint64_t>(segment_size)) * MB;
        }
        if (vm.count("name")) {    
            sdbf_sys.filename=(char*)input_name.c_str();
        }
        if (vm.count("index") && !vm.count("output")) {
            cerr << "sdhash:  ERROR: indexing requires output base filename " << endl;
            return -1;
        }
    }
    catch(exception& e)
    {
        cout << e.what() << "\n";
        return 0;
    }    
    // Initialization
    // set up sdbf object with current options
    sdbf::config = new sdbf_conf(sdbf_sys.thread_cnt, sdbf_sys.warnings, _MAX_ELEM_COUNT, _MAX_ELEM_COUNT_DD);
    // possible two sets to load for comparisons
    sdbf_set *set1 = new sdbf_set();
    sdbf_set *set2 = new sdbf_set();
    // indexing search support
    std::vector<bloom_filter *> indexlist;
    std::vector<sdbf_set *> setlist;
    index_info *info=(index_info*)malloc(sizeof(index_info));
    info->index=NULL;
    info->indexlist=&indexlist;
    info->setlist=&setlist;
    info->search_deep=true;
    if (vm.count("search-first")) {
        info->search_first=true;
        info->search_deep=true;
    } else
        info->search_first=false;
    if (vm.count("basename"))
        info->basename=true;
    else 
        info->basename=false;
    if (vm.count("index-dir")) {
	if (sdbf_sys.verbose)
	    cerr << "loading indexes ";
        if (fs::is_directory(idx_dir.c_str())) {
            for (fs::directory_iterator itr(idx_dir.c_str()); itr!=fs::directory_iterator(); ++itr) {
              if (fs::is_regular_file(itr->status()) && (itr->path().extension().string() == ".idx")) {
                 bloom_filter *indextest=new bloom_filter(itr->path().string());
                 if (sdbf_sys.verbose && (indexlist.size() % 5 == 0))
                     cerr<< "." ;
                 indexlist.push_back(indextest);
                 sdbf_set *tmp=new sdbf_set((idx_dir+"/"+itr->path().stem().string()).c_str());
                 setlist.push_back(tmp);
		 tmp->index=indextest;
              }
            }
        }
	if (sdbf_sys.verbose)
	    cerr << "done"<< endl;
    }

    // Perform all-pairs comparison
    if (vm.count("compare")) {
        if (inputlist.size()==1) {
            std::string resultlist;
            // load first set
            try {
                set1=new sdbf_set(inputlist[0].c_str());
            } catch (int e) {
                cerr << "sdhash: ERROR: Could not load SDBF file "<< inputlist[0] << ". Exiting"<< endl;
                return -1;
            }
            resultlist=set1->compare_all(sdbf_sys.output_threshold);
            cout << resultlist;
        } else if (inputlist.size()==2) {
            try {
                set1=new sdbf_set(inputlist[0].c_str());
            } catch (int e) {
                cerr << "sdhash: ERROR: Could not load SDBF file "<< inputlist[0] << ". Exiting"<< endl;
                return -1;
            }
            // load second set for comparison
            try {
                set2=new sdbf_set(inputlist[1].c_str());
            } catch (int e) {
                cerr << "sdhash: ERROR: Could not load SDBF file "<< inputlist[1] << ". Exiting"<< endl;
                return -1;
            }
            std::string resultlist;
            resultlist=set1->compare_to(set2,sdbf_sys.output_threshold, sdbf_sys.sample_size);
            cout << resultlist;
        } else  {
            cerr << "sdhash: ERROR: Comparison requires 1 or 2 arguments." << endl;
            delete set1;
            delete set2;
            return -1;
        }
        int n;
        if (set1!=NULL) {
            for (n=0;n< set1->size(); n++) 
                delete set1->at(n);
            delete set1;
        }
        if (set2!=NULL) {
            for (n=0;n< set2->size(); n++) 
                delete set2->at(n);
            delete set2;
        }
        return 0;
    }
    // Perform tow comparison
    if (vm.count("benchmark")) {
        if (inputlist.size()==2) {
            try {
                set1=new sdbf_set(inputlist[0].c_str());
            } catch (int e) {
                cerr << "sdhash: ERROR: Could not load SDBF file "<< inputlist[0] << ". Exiting"<< endl;
                return -1;
            }
            // load second set for comparison
            std::string resultlist;
            std::string against_sdbf_buffer = read_file(inputlist[1].c_str());
            double begin = utcsecond();
            int loop = 1000;
            for (int i = 0; i < loop; ++i) {
                set2=new sdbf_set(against_sdbf_buffer.data(), against_sdbf_buffer.size());
                resultlist=set1->compare_to(set2,sdbf_sys.output_threshold, sdbf_sys.sample_size);
                sdbf_set::destory(set2);
            }
            double end = utcsecond();
            cout << "cost=" << end - begin << " qps=" << loop/(end-begin) << " loop=" << loop << " result:" << resultlist;

            //sdbf_set *set2_cmp=new sdbf_set(inputlist[1].c_str());
            //resultlist=set1->compare_to(set2_cmp,sdbf_sys.output_threshold, sdbf_sys.sample_size);
            //cout << "result:" << resultlist;
            //cout << "new:set2:[" << set2 << "]\n";
            //cout << "old:set2:[" << set2_cmp << "]\n";
        } else  {
            cerr << "sdhash: ERROR: Comparison requires 1 or 2 arguments." << endl;
            delete set1;
            delete set2;
            return -1;
        }
        int n;
        if (set1!=NULL) {
            for (n=0;n< set1->size(); n++) 
                delete set1->at(n);
            delete set1;
        }
        if (set2!=NULL) {
            for (n=0;n< set2->size(); n++) 
                delete set2->at(n);
            delete set2;
        }
        return 0;
    }
    // validate hashes 
    if (vm.count("validate")) {
        for (i=0; i< inputlist.size(); i++) { 
            // load each set and throw it away
            if (!fs::is_regular_file(inputlist[i]))  {
                cout << "sdhash: ERROR file " << inputlist[i] << " not readable or not found." << endl;
                continue;
            }
            try {
                set1=new sdbf_set(inputlist[i].c_str());
                cout << "sdhash: file " << inputlist[i];
                cout << " SDBFs valid, contains " << set1->size() << " hashes." << endl;
                cout << "contains "<< set1->filter_count() << " bloom filters." << endl;
                cout << set1->input_size() << " total data size. " << endl;
            } catch (int e) {
                cerr << "sdhash: ERROR: Could not load file of SDBFs, "<< inputlist[i] << " is empty or invalid."<< endl;
                continue;
            }
            if (set1!=NULL) {
                for (int n=0;n< set1->size(); n++) 
                    delete set1->at(n);
                delete set1;
            }
        }
        return 0;
    }
    std::vector<string> small;
    std::vector<string> large;
    // Otherwise we are hashing. Make sure we have files.
    if (vm.count("input-files")) {
        // process stdin -- look for - arg
        if (inputlist.size()==1 && !inputlist[0].compare("-")) {
            if (sdbf_sys.segment_size == 0) {
                sdbf_sys.segment_size = 128*MB; // not currently allowing no segments
            }
            if (sdbf_sys.dd_block_size >  0) {
                set1=sdbf_hash_stdin(info);
            } 
            else { 
                // block size is always going to be defaulted in this case
                sdbf_sys.dd_block_size = 16;
                set1=sdbf_hash_stdin(info);
            }
        } else {
            // input list iterator
            if (sdbf_sys.verbose) 
                cerr << "sdhash: Building list of files to be hashed" << endl;
            vector<string>::iterator inp;
            for (inp=inputlist.begin(); inp < inputlist.end(); inp++) {
                // if recursive mode, then check directories as well
                if ( vm.count("deep") )  {
                    try {
                        if (fs::is_directory(*inp)) 
                            for ( fs::recursive_directory_iterator end, it(*inp); it!= end; ++it ) 
                                if (boost::filesystem::is_regular_file(*it)) {
                                    if (sdbf_sys.verbose) 
                                        cerr << "sdhash: adding file to hashlist "<< it->path().string() << endl;
                                    if (fs::file_size(*it) < 16*MB)
                                        small.push_back(it->path().string());
                                    else
                                        large.push_back(it->path().string());
                                    if ((fs::file_size(it->path()) >= sdbf_sys.segment_size) && sdbf_sys.warnings )  {
                                        cerr << "Warning: file " << it->path().string() << " will be segmented in ";
                                        cerr << sdbf_sys.segment_size/MB << "MB chunks prior to hashing."<< endl; 
                                    }
                                }
                    } catch (fs::filesystem_error err) {
                        cerr << "sdhash: ERROR: Filesystem problem in recursive searching " ;
                        cerr << err.what() << endl;
                        continue;
                    }
                } // always check if regular file. 
                if (fs::is_regular_file(*inp)) {    
                    if (sdbf_sys.verbose) 
                        cerr << "sdhash: adding file to hashlist "<< *inp << endl;
                    if (fs::file_size(*inp) < 16*MB) 
                        small.push_back(*inp);
                    else 
                        large.push_back(*inp);
                    if ((fs::file_size(*inp) >= sdbf_sys.segment_size) && sdbf_sys.warnings )  {
                        cerr << "sdhash: Warning: file " << *inp << " will be segmented in ";
                        cerr << sdbf_sys.segment_size/MB << "MB chunks prior to hashing."<< endl; 
                    }
                }
            }
        }
    } else if (vm.count("hash-list")) {
        // hash from a list in a file
        struct stat stat_res;
        if( stat( listingfile.c_str(), &stat_res) != 0) {
            cerr << "sdhash: ERROR: Could not access input file "<< listingfile<< ". Exiting." << endl;
            return -1;
        }
        processed_file_t *mlist=process_file(listingfile.c_str(), 1, sdbf_sys.warnings);
        if (!mlist) {
            cerr << "sdhash: ERROR: Could not access input file "<< listingfile<< ". Exiting." << endl;
            return -1;
        }
        i=0;
        std::istringstream fromfile((char*)mlist->buffer);
        std::string fname;
        while (std::getline(fromfile,fname)) {
            if (fs::is_regular_file(fname)) {
                if (fs::file_size(fname) < 16*MB) {
                    small.push_back(fname);
                } else {
                    large.push_back(fname);
                }    
                if ((fs::file_size(fname) >= sdbf_sys.segment_size) && sdbf_sys.warnings )  {
                    cerr << "sdhash: Warning: file " << fname << " will be segmented in ";
                    cerr << sdbf_sys.segment_size/MB << "MB chunks prior to hashing."<< endl; 
                }
            }
        }
    } else {
         cout << VERSION_INFO << ", rev " << REVISION << endl;
         cout << "       http://sdhash.org, license Apache v2.0" << endl;
         cout << "Usage: sdhash <options> <source files>|<hash files>"<< endl;
         cout << config << endl;
         return 0;
    }
    // Having built our lists of small/large files, hash them.
    int smallct=small.size();
    int largect=large.size();
    // from here, if we are indexing on creation, build things differently.
    if (vm.count("index")) {
        delete set1;
	int status = hash_index_stringlist(small,output_name);
	int status2 = hash_index_stringlist(large,output_name);
	return 0;
    } else {
        hash_start=time(0);         
        if (smallct > 0) {
            if (sdbf_sys.verbose)
                cerr << "sdhash: hashing small files"<< endl;
            char **smalllist=(char **)alloc_check(ALLOC_ONLY,smallct*sizeof(char*),"main", "filename list", ERROR_EXIT);
            for (i=0; i < smallct ; i++) {
                smalllist[i]=(char*)alloc_check(ALLOC_ONLY,small[i].length()+1, "main", "filename", ERROR_EXIT);
                strncpy(smalllist[i],small[i].c_str(),small[i].length()+1);
            }
            if (sdbf_sys.dd_block_size < 1 )  {
                if (vm.count("gen-compare") || vm.count("output")||vm.count("index-dir")) // if we need to save this set for comparison
                    sdbf_hash_files( smalllist, smallct, sdbf_sys.thread_cnt,set1, info);
                else 
                    sdbf_hash_files( smalllist, smallct, sdbf_sys.thread_cnt,NULL, info);
            } else {
                if (vm.count("gen-compare") || vm.count("output")||vm.count("index-dir"))
                    sdbf_hash_files_dd( smalllist, smallct, sdbf_sys.dd_block_size*KB,sdbf_sys.segment_size, set1, info);
		else 
                    sdbf_hash_files_dd( smalllist, smallct, sdbf_sys.dd_block_size*KB,sdbf_sys.segment_size, NULL, info);
            }
        }
        if (largect > 0) {
            if (sdbf_sys.verbose)
                cerr << "sdhash: hashing large files"<< endl;
            char **largelist=(char **)alloc_check(ALLOC_ONLY,largect*sizeof(char*),"main", "filename list", ERROR_EXIT);
            for (i=0; i < largect ; i++) {
                largelist[i]=(char*)alloc_check(ALLOC_ONLY,large[i].length()+1, "main", "filename", ERROR_EXIT);
                strncpy(largelist[i],large[i].c_str(),large[i].length()+1);
            }
            if (sdbf_sys.dd_block_size == 0 ) {
                if (vm.count("gen-compare")) // if we need to save this set for comparison
                    sdbf_hash_files( largelist, largect, sdbf_sys.thread_cnt,set1, info);
                else
                    sdbf_hash_files( largelist, largect, sdbf_sys.thread_cnt,NULL, info);
            } else {
                if (sdbf_sys.dd_block_size == -1) { 
                    if (sdbf_sys.warnings || sdbf_sys.verbose) 
                       cerr << "sdhash: Warning: files over 16MB are being hashed in block mode. Use -b 0 to disable." << endl;
                    if (vm.count("gen-compare")|| vm.count("output") ||vm.count("index-dir")) // if we need to save this set for comparison
                        sdbf_hash_files_dd( largelist, largect, 16*KB,sdbf_sys.segment_size, set1, info);
                    else
                        sdbf_hash_files_dd( largelist, largect, 16*KB,sdbf_sys.segment_size, NULL, info);
                } else {
                    if (vm.count("gen-compare")||vm.count("output") ||vm.count("index-dir")) // if we need to save this set for comparison
                        sdbf_hash_files_dd( largelist, largect, sdbf_sys.dd_block_size*KB,sdbf_sys.segment_size, set1, info);
                    else 
                        sdbf_hash_files_dd( largelist, largect, sdbf_sys.dd_block_size*KB,sdbf_sys.segment_size, NULL, info);
                }
            }
        } // if large files exist
        hash_end=time(0);
        if (sdbf_sys.verbose)
            cerr << hash_end - hash_start << " seconds hash time" << endl;
    } // if not indexing
    // print it out if we've been asked to
    if (vm.count("gen-compare")) {
        string resultlist;
        resultlist=set1->compare_all(sdbf_sys.output_threshold);
        cout << resultlist;
    } else {
        if (vm.count("output")) {
            if (vm.count("index-dir")) {
                string output_indexr = output_name + ".idx-result";
		std::filebuf fb;
                fb.open (output_indexr.c_str(),ios::out|ios::binary);
                if (fb.is_open()) {
                    std::ostream os(&fb);
                    os << set1->index_results();
                    fb.close();
                } else {
                    cerr << "sdhash: ERROR cannot write to file " << output_indexr<< endl;
                    return -1;
                }
            } else { // search-index is destructive, so we do not make actual sdhashes in this case
		output_name= output_name+".sdbf";
		std::filebuf fb;
		fb.open (output_name.c_str(),ios::out|ios::binary);
		if (fb.is_open()) {
		    std::ostream os(&fb);
		    os << set1;
		    fb.close();
		} else {
		    cerr << "sdhash: ERROR cannot write to file " << output_name<< endl;
		    return -1;
		}
 	    }
        } else {
            if (vm.count("index-dir")) 
                cerr << set1->index_results();
	    else 
                cout << set1;
        }
    }
    if (setlist.size() > 0) {
        for (int n=0;n< setlist.size(); n++) {
            for (int m=0;m< setlist.at(n)->size(); m++) {
               delete setlist.at(n)->at(m);
	    }
	    delete setlist.at(n)->index;
	    delete setlist.at(n);
	}
    }
    if (set1!=NULL) {
        for (int n=0;n< set1->size(); n++) 
            delete set1->at(n);
        delete set1;
    }
    if (set2!=NULL) {
        for (int n=0;n< set2->size(); n++) 
            delete set2->at(n);
        delete set2;
    }
    if (info)
	free(info);
    return 0;
}
